### ISSUE TITLE
should we normalize scheme and hostname of URI to be lower?

### ISSUE THREAD

**Issue by dstandish** [apache/airflow#25186]
datasets are defined by their URI

we don't really spelly out what exactly our URI standard is, apart from reserving the `airflow` scheme and  ascii

for websites, the scheme and hostname are case insensitive, while everything else is not.

should we normalize scheme and hostname or allow case sensitive differentiation?

my inclination is, we should allow everything to be case sensitive, except _possibly_ the scheme.  and the reason is, we don't know exactly what "hostname" will mean for a dataset.  if, for example,  it's a database object, it could be case sensitive.  scheme.

if we do implement some normalization, we introduce somewhat of a problem when we do sqlalchemyf lookups by URI because we can't just do `Dataset.uri == uri`; we'd have to normalize the incoming URI value first.  One possibility is to split out the scheme into a different column in the db that has CI collation, which would avoid this issue, though it would force you to decompose.  but this is messy too.

**Comment by potiuk** [comment]
I believe we should just follow the standards (also because if anyone would like to use any tools to compare the URIs for example might use a tool that is RFC3986 compliant:

This is the URI's RFC: https://datatracker.ietf.org/doc/html/rfc3986 (not this is not HTTP URL RFC it's generir URI RFC so very much our case)

*Although schemes are case-insensitive, the canonical form is lowercase and documents that specify schemes must do so with lowercase letters
* The host subcomponent is case-insensitive
* Although host is case-insensitive, producers and normalizers should use lowercase for registered names and hexadecimal addresses for the sake ofuniformity, while only using uppercase letters for percent-encodings.


```
6.2.2.1.  Case Normalization

   For all URIs, the hexadecimal digits within a percent-encoding
   triplet (e.g., "%3a" versus "%3A") are case-insensitive and therefore
   should be normalized to use uppercase letters for the digits A-F.

   When a URI uses components of the generic syntax, the component
   syntax equivalence rules always apply; namely, that the scheme and
   host are case-insensitive and therefore should be normalized to
   lowercase.  For example, the URI <HTTP://www.EXAMPLE.com/> is
   equivalent to <http://www.example.com/>.  The other generic syntax
   components are assumed to be case-sensitive unless specifically
   defined otherwise by the scheme (see Section 6.2.3).
```

The rest is case-sensitive by default. We can define some specific parts of the URI case insensitive but it dependes on scheme:

```
   Some schemes define additional subcomponents that consist of case-
   insensitive data, giving an implicit license to normalizers to
   convert this data to a common case (e.g., all lowercase).  For
   example, URI schemes that define a subcomponent of path to contain an
   Internet hostname, such as the "mailto" URI scheme, cause that
   subcomponent to be case-insensitive and thus subject to case
   normalization (e.g., "mailto:Joe@Example.COM" is equivalent to
   "mailto:Joe@example.com", even though the generic syntax considers
   the path component to be case-sensitive).
```

I think we do not have much freedom there in terms of host being case-sensitive. It cannot be.

**Comment by dstandish** [comment]
> I think we do not have much freedom there in terms of host being case-sensitive. It cannot be.

What about Airflow's Connection URI?  We do not do any such normalization there.

**Comment by potiuk** [comment]
> What about Airflow's Connection URI? We do not do any such normalization there. And we still call it a URI.

We do. We use urlparse which is RFC3986 compliant (and will do the normalization for us). https://docs.python.org/3/library/urllib.parse.html

**Comment by potiuk** [comment]
(It even has built in support for a number of schemas that implemented their own case-insensitive components)

**Comment by dstandish** [comment]
it doesn't though!

```
>>> from urllib.parse import urlparse
>>> urlparse('s3://Abc') == urlparse('s3://abc')
False
>>> urlparse('s3://abc') == urlparse('s3://abc')
True
```

**Comment by dstandish** [comment]
oh interesting:
```
urlparse('s3://abc').hostname
'abc'
urlparse('s3://Abc').hostname
'abc'
```

**Comment by potiuk** [comment]
Yep. This is what I was referring to - I believe I even fixed a bug some 2 years ago about RFC3986 compliance in Airflow. Can't find it quickly though

**Comment by dstandish** [comment]
so if you call the hostname then it renders as lower.  but maybe by doing that you are implicitly agreeing to that standard.

anyway we don't use this in airflow conn uri parsing.  so we do _not_ normalize in this way and, in effect, treat it as case sensitive.

here it seems we pretty intentionally avoid use of urlparse.hostname for case-sensitivity reasons https://github.com/apache/airflow/blob/main/airflow/models/connection.py#L47-L50

**Comment by dstandish** [comment]
yeah i think that line is probably what you're referring to ⬆️

**Comment by dstandish** [comment]
here's the PR: https://github.com/apache/airflow/pull/4591

**Comment by potiuk** [comment]
Yep. It's a mess. But I think if we are about to  be compliant with OpenLineage and others, being compliant with RFC. The worst we can do is invent yet another "our" interpretation of the standard:

![image](https://user-images.githubusercontent.com/595491/180037810-062285f2-5ade-4d28-8bb2-21475de667f7.png)

**Comment by dstandish** [comment]
yeah you are probably right.  i'm not saying that we necessarily create a new standard.  what's at stake for me is whether airflow takes on the responsibility of normalizing e.g. when storing in the database.  in other words, given that the URI field is the unique identifier for an airflow dataset, does airflow really need to take on the responsibility of normalizing, such that when a user's code handles it in two different ways, we treat it  as the same dataset.  i'm reluctant to take on that responsibility, and i am more inclined to force the user to just make their code consistent.  in other words, just store in a fully case-sensitivie collation, and not trouble ourselves with, e.g. decomposing, lowering hostname, recomposing and storing -- or, alternatively, storing hostname separately in a case-insensitive field and merge back in on read etc.  i would rather not.

but i think we probably need to decide before 2.4 because a change to this behavior would be breaking.

unless we want to mark datasets as experimental... which i doubt...  i digress

**Comment by potiuk** [comment]
I think we should at least detect "fat finger" problems. I.e when somoene *Inside airflow installation* creates two different datasets with equivalent urls, we should not allow that. We are able to do do that very easily and warn the user. 

I am perfectly ok with storing dataset with URL without normalisation. But at least we should have an unique index which will prevent the user from creating two different datasets with two equivalent (but different) URIs. This is not difficult. We can for example fully normalize the URI, convert into base64-encoded string and save it as "unique_id" or smth in the database. Then whenever we are inserting a dataset with different URI and same "unique_id", we simply fail with:
"This URI here is the same as that URI there".  Shoudl not be very complex, and I think it prevents users from making silly errors that will be difficult to debug otherwise.

There is no real drawback of it that I can think except 'generate_unique_id" using normalisation. Pretty much no performance penaly, much better user experience. Airflow helping user to make less mistakes.

**Comment by dstandish** [comment]
I am actually thinking now that we should abandon URI as the identifier for datasets.  I think part of the motivation for choosing URI was [this document in openlineage](https://github.com/OpenLineage/OpenLineage/blob/main/spec/Naming.md).  But the thing is, openlineage doesn't actually use URI for datasets!  The document is misleading.  In OpenLineage, what defines a dataset is actually `namespace + name`.  To me, this is much better and I think we should change Dataset to do something more similar to that -- if not exactly that.

As we've seen, URI is problematic for a number of reasons.  Just to name a few... It isn't "uniquely" defined by a string --- there are _many_ such URI that correspond to the same exact resource.  It's not human friendly, and really requires that you _also_ have a friendly _name_ you can refer to it with (e.g. to present in UI).  It can be confused for "connection string" -- when we're mearely looking for a label for a dataset.  URI makes more sense as something generated by machines -- not as the human-entered label for a dataset.  But openlineage has no need for it anyway.  I plan to look at some other lineage standards to see how they identify datasets.  But for now I am really feeling that we shouldn't use URI for this purpose.

**Comment by potiuk** [comment]
100% on board.

**Comment by dstandish** [comment]
lost the battle on this one @potiuk.  still URI.

**Comment by potiuk** [comment]
Now we need to decide on what is the acceptable false-positive on duplicated URIs we can go with. This will hit us sooner or later, so it's quite necessary to make a conscious decision on it and accept the consequences, rather than leave it for interpretation.

**Comment by dstandish** [comment]
er false negative, i think you mean.

**Comment by potiuk** [comment]
positive :)

**Comment by potiuk** [comment]
I mean positive that it is false-negative, that is :D

**Comment by dstandish** [comment]
Haha nice

**Comment by eladkal** [comment]
Discussed this with @uranusjr and @cmarteepants Airflow 3 is great opportunity to set lowercase

**Comment by uranusjr** [comment]
We never got aroud to actually do this. Is it too late? I hope we could treat this as a bug fix (?)
